twine/package.py

```diff
@@ -18,39 +18,35 @@
 import os
 import re
 import subprocess
-import sys
-import warnings
-from typing import (
-    Any,
-    Dict,
-    Iterable,
-    List,
-    NamedTuple,
-    Optional,
-    Sequence,
-    Tuple,
-    Union,
-    cast,
-)
-
-if sys.version_info >= (3, 10):
-    import importlib.metadata as importlib_metadata
-else:
-    import importlib_metadata
-
-import packaging.version
-import pkginfo
+from typing import Any, Dict, List, NamedTuple, Optional, Tuple
+
+from packaging import metadata
+from packaging import version
 from rich import print
 
+from twine import bdist
 from twine import exceptions
+from twine import sdist
 from twine import wheel
 from twine import wininst
 
+# Monkeypatch Metadata 2.0 support
+metadata._VALID_METADATA_VERSIONS = [
+    "1.0",
+    "1.1",
+    "1.2",
+    "2.0",
+    "2.1",
+    "2.2",
+    "2.3",
+    "2.4",
+]
+
 DIST_TYPES = {
     "bdist_wheel": wheel.Wheel,
     "bdist_wininst": wininst.WinInst,
-    "bdist_egg": pkginfo.BDist,
-    "sdist": pkginfo.SDist,
+    "bdist_egg": bdist.BDist,
+    "sdist": sdist.SDist,
```

Comment:
I'm uneasy replacing an established implementation with an inline implementation of something very similar. Can you point to the rationale for re-implementing parts of pkginfo? My preference is to converge implementations and not fork them.

Reply:
I can maybe use the implementation in `pkginfo` but to me it seems unnecessary to have a dependency to an external package to extract a member from a zip or tar archive. `twine` already implements it for two distribution types. Also, I have the impression that the `pkginfo.Distribution.read()` method is an implementation detail.